Sources/SwiftInspectorVisitors/TypeDescription.swift

```diff
@@ -0,0 +1,238 @@
+      self = .member(name: text, baseType: baseType)
+
+    } else if caseDescription == Self.optionalDescription {
+      let typeDescription = try values.decode(Self.self, forKey: .typeDescription)
+      self = .optional(typeDescription)
+
+    } else if caseDescription == Self.implicitlyUnwrappedDescription {
+      let typeDescription = try values.decode(Self.self, forKey: .typeDescription)
+      self = .implicitlyUnwrapped(typeDescription)
+
+    } else if caseDescription == Self.arrayDescription {
+      let typeDescription = try values.decode(Self.self, forKey: .typeDescription)
+      self = .array(typeDescription)
+
+    } else if caseDescription == Self.dictionaryDescription {
+      let key = try values.decode(Self.self, forKey: .typeDescriptionKey)
+      let value = try values.decode(Self.self, forKey: .typeDescriptionValue)
+      self = .dictionary(key: key, value: value)
+
+    } else if caseDescription == Self.compositionDescription {
+      let typeDescriptions = try values.decode([Self].self, forKey: .typeDescriptions)
+      self = .composition(typeDescriptions)
+
+    } else if caseDescription == Self.tupleDescription {
+      let typeDescriptions = try values.decode([Self].self, forKey: .typeDescriptions)
+      self = .tuple(typeDescriptions)
+
+    } else {
+      throw CodingError.unknownCase
+    }
+  }
+
+  public func encode(to encoder: Encoder) throws {
+    var container = encoder.container(keyedBy: CodingKeys.self)
+    try container.encode(caseDescription, forKey: .caseDescription)
+    switch self {
+    case let .simple(name):
+      try container.encode(name, forKey: .text)
+    case let .unknown(text):
+      try container.encode(text, forKey: .text)
+    case let .optional(type),
+         let .array(type),
+         let .implicitlyUnwrapped(type):
+      try container.encode(type, forKey: .typeDescription)
+    case let .tuple(types),
+         let .composition(types):
+      try container.encode(types, forKey: .typeDescriptions)
+    case let .dictionary(key, value):
+      try container.encode(key, forKey: .typeDescriptionKey)
+      try container.encode(value, forKey: .typeDescriptionValue)
+    case let .member(name, baseType):
+      try container.encode(name, forKey: .text)
+      try container.encode(baseType, forKey: .typeDescription)
+    }
+  }
+
+  enum CodingKeys: String, CodingKey {
+    case caseDescription
+    case text
+    case typeDescription
+    case typeDescriptions
+    case typeDescriptionKey
+    case typeDescriptionValue
+  }
+
+  enum CodingError: Error {
+    case unknownCase
+  }
+
+  /// - Parameters:
+  ///   - name: The simple name of the returned type.
+  ///   - parent: The parent type for the returned type.
+  /// - Returns: Returns a type description of case `.member` with the given name as the name and the receiver as the base type.
+  /// - Note: This method only makes sense when the `parent` is of case  `simple`, `member`, `optional`, and `implicitlyUnwrapped`.
```

Comment:
Is there a way we can enforce this?

Reply:
We could `switch` and then assert internally... but I'm not sure that's worth it just yet. No way to compile-time enforce afaik.